Lib/ufo2ft/featureWriters/gdefFeatureWriter.py

```diff
@@ -0,0 +1,160 @@
+from ufo2ft.constants import OPENTYPE_CATEGORIES_KEY
+from ufo2ft.featureWriters import BaseFeatureWriter, ast
+
+
+class GdefFeatureWriter(BaseFeatureWriter):
+    """Generates a GDEF table based on OpenType Category and glyph anchors.
+
+    There are a few use cases:
+        1. There is no GDEF data in the UFO.
+        2. There is some GDEF data in the UFO, in the UFO lib
+           'public.openTypeCategories' or with ligature caret anchors
+           'caret_<number>'.
+        3. There is a GDEF in the features, and some GDEF data.
+
+    This first generates a GDEF from the features, from the GDEF table
+    definition if present or from the lookups.
+    Then it updates the GDEF with the public.openTypeCategories and the
+    ligature anchors for glyphs that are not already in GlyphClassDefs or
+    do not have ligature carets.
```

Comment:
Is there a way to say this glyph has no GDEF class but please don’t auto-generate one, would setting `public.openTypeCategories` to an empty string signal that?

Reply:
No, `public.openTypeCategories` values can only be `base`, `ligature`, `mark`, `component`.

If no GDEF is defined in the feature code and the glyph is not in `public.openTypeCategories`, it will only be in a GDEF Glyph Class if a lookup type makes it so.

Reply:
OK.

Reply:
https://unifiedfontobject.org/versions/ufo3/lib.plist/#publicopentypecategories

> Values must be one of base, mark, ligature or component.

So you can only leave it out, but then teh writer would assume `auto` :o

Reply:
> but then the writer would assume auto :o

Which is exactly what I’m trying to avoid. I’m converting fonts from SFD where you can leave out glyph class while opting out from automatic behavior. I currently also generate a GDEF table, but I was checking if I can skip that bit now.

Reply:
Could you describe how glyphs end up in GDEF Glyphs Class when you don't want to? What's the actual use case?

Reply:
I have no specific use case, if the SFD says this glyph has no GDEF class and don’t guess one, I want to maintain that. Say I have a place holder glyph that gets decomposed in `ccmp` feature but uses components, it might get anchors propagated to it and classified as base glyph, but GDEF class does not matter for it and it would save a few bytes to not et a GDEF glyph class to it (and if you have a few hundreds of these glyphs it might add up).